File: leet/InvertedBinaryTreeLevelOrderTraversal.java
import java.util.ArrayList;
import java.util.Collections;
import java.util.LinkedList;
import java.util.List;
import java.util.Queue;
import java.util.Stack;

/**
 * Created by rishabh on 9/14/2016.
 */
public class InvertedBinaryTreeLevelOrderTraversal {
    public static void main(String[] args) {
        InvertedBinaryTreeLevelOrderTraversal traversal = new InvertedBinaryTreeLevelOrderTraversal();
        BinaryTree root = new BinaryTree(1);
        root.left = new BinaryTree(2);
        root.right = new BinaryTree(3);
        root.left.left = new BinaryTree(4);
        root.left.right = new BinaryTree(5);
        root.right.left = new BinaryTree(6);
        root.right.right = new BinaryTree(7);
        List<List<Integer>> order = traversal.levelOrderBottom(root);
//        List<List<Integer>> order = traversal.levelOrderBottom(null);
        for (List<Integer> list : order) {
            for (int num : list) {
                System.out.print(num + "\t");
            }
            System.out.println();
        }
    }
    private List<List<Integer>> levelOrderBottom(BinaryTree root) {
        if (root == null) {
            return Collections.EMPTY_LIST;
        }
        Queue<BinaryTree> orderedQueue = new LinkedList<>();
        Queue<BinaryTree> auxQueue = new LinkedList<>();
        orderedQueue.add(root);
        Stack<List<Integer>> inverseStack = new Stack<>();
        List<Integer> levelList = new ArrayList<>();
        while (!orderedQueue.isEmpty()) {
            BinaryTree current = orderedQueue.remove();
            levelList.add(current.val);
            if(current.left != null) {
                auxQueue.add(current.left);
            }
            if(current.right != null) {
                auxQueue.add(current.right);
            }
            if(orderedQueue.isEmpty()) {
                while (!auxQueue.isEmpty()) {
                    orderedQueue.add(auxQueue.remove());
                }
                inverseStack.add(levelList);
                levelList = new ArrayList<>();
            }
        }
        List<List<Integer>> result = new ArrayList<>();
        while (!inverseStack.isEmpty()) {
            result.add(inverseStack.pop());
        }
        return result;
    }
}

File: leet/InvertedBinaryTreeLevelOrderTraversal.py
class TreeNode(object):
    def __init__(self, x):
        self.val = x
        self.left = None
        self.right = None


class Solution(object):
    def levelOrderBottom(self, root):
        """
        :type root: TreeNode
        :rtype: List[List[int]]
        """
        if root is None:
            return []
        from collections import deque
        orderedQueue = deque()
        auxQueue = deque()
        orderedQueue.append(root)
        inverseStack = list()
        levelList = list()
        while len(orderedQueue) is not 0:
            current = orderedQueue.popleft()
            levelList.append(current.val)
            if current.left is not None:
                auxQueue.append(current.left)
            if current.right is not None:
                auxQueue.append(current.right)
            if len(orderedQueue) is 0:
                while len(auxQueue) is not 0:
                    orderedQueue.append(auxQueue.popleft())
                inverseStack.append(levelList)
                levelList = list()
        result = list()
        while len(inverseStack) is not 0:
            result.append(inverseStack.pop())
        return result


soln = Solution()
root = TreeNode(1)
root.left = TreeNode(2)
root.right = TreeNode(3)
root.left.left = TreeNode(4)
root.left.right = TreeNode(5)
root.right.left = TreeNode(6)
root.right.right = TreeNode(7)

print(soln.levelOrderBottom(root))
